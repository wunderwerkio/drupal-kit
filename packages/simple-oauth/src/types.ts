export interface SimpleOauthTokenResponse {
  token_type: "Bearer";
  expires_in: number;
  access_token: string;
  refresh_token: string;
}

export interface SimpleOauthErrorResponse {
  error: SimpleOauthError;
  error_description?: string;
  hint?: string;
}

export interface SimpleOauthGrant {
  client_id: string;
  client_secret: string;
}

export interface SimpleOauthAuthCodeGrant extends SimpleOauthGrant {
  code: string;
  redirect_uri?: string;
}

export interface SimpleOauthClientCredentialsGrant extends SimpleOauthGrant {
  scope?: string;
}

export interface SimpleOauthRefreshTokenGrant extends SimpleOauthGrant {
  refresh_token: string;
  scope?: string;
}

export interface SimpleOauthPasswordGrant extends SimpleOauthGrant {
  username: string;
  password: string;
  scope?: string;
}

/**
 * Augment this interface if you want to implement your
 * own OAuth grant types.
 */
export interface SimpleOauthGrantTypes {
  authorization_code: SimpleOauthAuthCodeGrant;
  client_credentials: SimpleOauthClientCredentialsGrant;
  refresh_token: SimpleOauthRefreshTokenGrant;
  password: SimpleOauthPasswordGrant;
}

export type SimpleOauthError =
  | "invalid_request"
  | "unsupported_grant_type"
  | "invalid_client"
  | "invalid_scope"
  | "server_error"
  | "access_denied"
  | "invalid_grant";

export type SimpleOauthInvalidRequest =
  | "auth_code_expired"
  | "auth_code_malformed"
  | "auth_code_revoked"
  | "auth_code_decrypt_error"
  | "auth_code_wrong_client"
  | "redirect_uri_invalid"
  | "code_verifier_rfc_error"
  | "code_challenge_invalid"
  | "code_challenge_rfc_error"
  | "code_challenge_missing_public_client"
  | "generic"
  | "invalid_parameter_scope"
  | "invalid_parameter_client_secret"
  | "invalid_parameter_client_id"
  | "invalid_parameter_redirect_uri"
  | "invalid_parameter_refresh_token"
  | "invalid_parameter_state"
  | "invalid_parameter_redirect_uri"
  | "invalid_parameter_password"
  | "invalid_parameter_username"
  | "invalid_parameter_code_verifier"
  | "invalid_parameter_code";
