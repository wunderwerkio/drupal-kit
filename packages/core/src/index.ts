export * from "./Drupalkit";
export * from "./types";
