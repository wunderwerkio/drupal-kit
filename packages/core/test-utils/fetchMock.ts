import type { FetchMock } from "jest-fetch-mock";

import type { Drupalkit } from "../src";

export const mockResponse = (
  fm: FetchMock,
  dk: Drupalkit,
  options: {
    url: string | RegExp;
    status?: number;
    payloadFixture?: object;
    locale?: string;
    contentType?: string;
  },
) => {
  const url =
    typeof options.url === "string"
      ? dk.buildUrl(options.url, {
          localeOverride: options.locale,
        })
      : options.url;

  // Mock next request for url.
  fm.mockOnceIf(url, async (req) => {
    const payload = options.payloadFixture;

    return {
      body: payload ? JSON.stringify(payload) : "{}",
      headers: {
        "content-type": options.contentType ?? "application/json",
      },
      url: req.url,
      status: options.status ?? 200,
    };
  });
};

export const mockNetworkError = (fm: FetchMock) => {
  fm.mockRejectOnce(new Error("Network Error"));
};
