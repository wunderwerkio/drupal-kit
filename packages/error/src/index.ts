import * as DrupalkitTypes from "@drupalkit/types";

import { RequestErrorOptions as DrupalkitErrorOptions } from "./types";

/**
 * Custom error class to help with error handling.
 */
export class DrupalkitError extends Error {
  /**
   * Error name.
   */
  name: "HttpError";

  /**
   * HTTP status code.
   */
  status: number;

  /**
   * Request options that lead to the error.
   */
  request: DrupalkitTypes.RequestOptions;

  /**
   * Response object if a response was received.
   */
  response?: DrupalkitTypes.DrupalkitResponse<unknown>;

  /**
   * Construct a new DrupalkitError.
   *
   * @param message - Error message.
   * @param statusCode - HTTP status code.
   * @param options - Request / Response that lead to the error.
   */
  constructor(
    message: string,
    statusCode: number,
    options: DrupalkitErrorOptions,
  ) {
    super(message);

    // Maintains proper stack trace (only available on V8)
    // @ts-ignore
    if (Error.captureStackTrace) {
      // @ts-ignore
      Error.captureStackTrace(this, this.constructor);
    }

    this.name = "HttpError";
    this.status = statusCode;

    if (options.response) {
      this.response = options.response;
    }

    // redact request credentials without mutating original request options
    const requestCopy = Object.assign({}, options.request);
    if (options.request.headers.authorization) {
      requestCopy.headers = Object.assign({}, options.request.headers, {
        authorization: options.request.headers.authorization.replace(
          / .*$/,
          " [REDACTED]",
        ),
      });
    }

    this.request = requestCopy;
  }
}
