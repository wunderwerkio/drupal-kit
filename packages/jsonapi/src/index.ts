export * from "./resources";
export * from "./DrupalkitJsonApi";
export * from "./DrupalkitJsonApiError";
