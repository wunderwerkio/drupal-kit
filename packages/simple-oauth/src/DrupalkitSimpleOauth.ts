import { Ok } from "ts-results";
import { Drupalkit, DrupalkitOptions } from "@drupal-kit/core";

import { SimpleOauthGrantTypes, SimpleOauthTokenResponse } from "./types";

declare module "@drupal-kit/core" {
  interface DrupalkitOptions {
    oauthTokenEndpoint?: string;
  }
}

/**
 * DrupalkitSimpleOauth plugin for Drupalkit.
 *
 * Extends the Drupalkit instance to support requesting
 * tokens via the `simple_oauth` drupal module.
 *
 * @param drupalkit - The Drupalkit instance.
 * @param drupalkitOptions - The options for the Drupalkit instance.
 */
export const DrupalkitSimpleOauth = (
  drupalkit: Drupalkit,
  drupalkitOptions: DrupalkitOptions,
) => {
  /**
   * Extend the Drupalkit instance.
   */
  return {
    simpleOauth: {
      async requestToken<
        GrantType extends keyof SimpleOauthGrantTypes,
        Grant extends SimpleOauthGrantTypes[GrantType],
      >(grantType: GrantType, grant: Grant) {
        const path = drupalkitOptions.oauthTokenEndpoint ?? "/oauth/token";
        const url = drupalkit.buildUrl(path);

        const body = new URLSearchParams();
        body.append("grant_type", grantType);
        for (const [key, value] of Object.entries(grant)) {
          body.append(key, value);
        }

        const result = await drupalkit.request<SimpleOauthTokenResponse>(url, {
          method: "POST",
          body,
          headers: {
            "content-type": "application/x-www-form-urlencoded",
          },
        });

        if (result.err) {
          return result;
        }

        return Ok(result.val.data);
      },
    },
  };
};
