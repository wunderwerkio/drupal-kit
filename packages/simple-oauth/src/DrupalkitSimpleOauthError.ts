import { DrupalkitError, DrupalkitErrorOptions } from "@drupal-kit/core";

import { SimpleOauthError, SimpleOauthInvalidRequest } from "./types.js";

const InvalidRequestHintMap = {
  auth_code_revoked: "Authorization code has been revoked",
  auth_code_malformed: "Authorization code malformed",
  auth_code_expired: "Authorization code has expired",
  auth_code_wrong_client: "Authorization code was not issued to this client",
  auth_code_decrypt_error: "Cannot decrypt the authorization code",
  code_verifier_rfc_error:
    "Code Verifier must follow the specifications of RFC-7636.",
  redirect_uri_invalid: "Invalid redirect URI",
  code_challenge_invalid: /^Code challenge method must be one of .*$/,
  code_challenge_rfc_error:
    "Code challenge must follow the specifications of RFC-7636.",
  code_challenge_missing_public_client:
    "Code challenge must be provided for public clients",
};

/**
 * Enhanced DrupalkitError with SimpleOauth specific data.
 */
export class DrupalkitSimpleOauthError extends DrupalkitError {
  readonly error: SimpleOauthError;
  readonly hint?: string;

  /**
   * @inheritdoc
   */
  constructor(
    message: string,
    statusCode: number,
    error: SimpleOauthError,
    options: DrupalkitErrorOptions,
    hint?: string,
  ) {
    super(message, statusCode, options);

    this.error = error;
    this.hint = hint;
  }

  /**
   * Gets the specific type of invalid_request error.
   */
  public getInvalidRequestType(): SimpleOauthInvalidRequest {
    if (this.error !== "invalid_request" || !this.hint) {
      return "generic";
    }

    // Check invalid parameter hints.
    if (this.hint.includes("Check the")) {
      const regexp = /`(\w+)`/g;
      const result = regexp.exec(this.hint);
      if (result && result.length > 1) {
        return ("invalid_parameter_" + result[1]) as SimpleOauthInvalidRequest;
      }
    }

    // Check invalid request hints.
    for (const key of Object.keys(
      InvalidRequestHintMap,
    ) as (keyof typeof InvalidRequestHintMap)[]) {
      const value = InvalidRequestHintMap[key];

      if (typeof value === "string" && this.hint.includes(value)) {
        return key;
      }

      if (value instanceof RegExp) {
        if (this.hint.match(value)) {
          return key;
        }
      }
    }

    return "generic";
  }

  /**
   * Create new error instance from a DrupalkitError.
   *
   * @param error - The DrupalkitError to create an instance from.
   */
  public static fromDrupalkitError(error: DrupalkitError) {
    const errorData = error.response
      ? this.extractErrorFromResponse(error.response.data)
      : null;

    if (!errorData) {
      return error;
    }

    return new this(
      errorData.message ?? error.message,
      error.statusCode,
      errorData.error,
      {
        request: error.request,
        response: error.response,
      },
      errorData.hint,
    );
  }

  /**
   * Extract SimpleOauth error from response data.
   *
   * @param data - The response data to extract the error from.
   */
  public static extractErrorFromResponse(
    /* eslint-disable-next-line @typescript-eslint/no-explicit-any */
    data: any,
  ) {
    if (!data?.error) {
      return null;
    }

    return {
      error: data.error as SimpleOauthError,
      message: data.error_description as string | undefined,
      hint: data.hint as string | undefined,
    };
  }
}
