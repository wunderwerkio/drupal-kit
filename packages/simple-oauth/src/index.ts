import { DrupalkitSimpleOauth } from "./DrupalkitSimpleOauth";

export default DrupalkitSimpleOauth;
export { DrupalkitSimpleOauth };

export * from "./types";
export * from "./DrupalkitSimpleOauthError";
