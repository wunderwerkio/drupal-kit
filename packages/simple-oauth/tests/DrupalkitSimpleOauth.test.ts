import fetchMock, { enableFetchMocks } from "jest-fetch-mock";
import { Drupalkit, DrupalkitOptions } from "@drupal-kit/core";
import { mockNetworkError, mockResponse } from "@drupal-kit/core/test-utils";

import { DrupalkitSimpleOauth } from "../src/index";
import ErrorResponse from "./fixtures/error_response.json";
import TokenResponse from "./fixtures/token_response.json";

enableFetchMocks();

describe("DrupalkitSimpleOauth", () => {
  const BASE_URL = "https://my-drupal.com";

  const CLIENT_ID = "12345678901234567890123456789012";
  const CLIENT_SECRET = "F9w1cM0GQw7GjjQUaZcscWHtxnMOvn4d";

  const createDrupalkit = (
    options: DrupalkitOptions = {
      baseUrl: BASE_URL,
    },
  ) => {
    const EnhancedDrupalkit = Drupalkit.plugin(DrupalkitSimpleOauth);

    return new EnhancedDrupalkit({
      locale: "de",
      defaultLocale: "de",
      ...options,
    });
  };

  beforeEach(() => {
    fetchMock.resetMocks();
  });

  it("should instanciate", () => {
    const drupalkit = createDrupalkit();

    expect(drupalkit).toHaveProperty("simpleOauth");
  });

  it("should request token with client credentials grant", async () => {
    const drupalkit = createDrupalkit();

    mockResponse(fetchMock, drupalkit, {
      url: "/oauth/token",
      payloadFixture: TokenResponse,
    });

    const result = await drupalkit.simpleOauth.requestToken(
      "client_credentials",
      {
        client_id: CLIENT_ID,
        client_secret: CLIENT_SECRET,
      },
    );

    expect(result.ok).toBeTruthy();

    if (result.ok) {
      expect(result.val).toMatchSnapshot("token-response");
    }

    // @ts-ignore
    expect(fetchMock.mock.calls[0][1]).toMatchObject({
      method: "POST",
      headers: {
        "content-type": "application/x-www-form-urlencoded",
      },
    });

    // @ts-ignore
    expect(fetchMock.mock.calls[0][1].body).toBeInstanceOf(URLSearchParams);

    // @ts-ignore
    const body = fetchMock.mock.calls[0][1].body as URLSearchParams;
    expect(body.toString()).toMatchSnapshot("request-payload");
  });

  it("should request token explicit endpoint", async () => {
    const drupalkit = createDrupalkit({
      baseUrl: BASE_URL,
      oauthTokenEndpoint: "/custom/token",
    });

    mockResponse(fetchMock, drupalkit, {
      url: "/custom/token",
      payloadFixture: TokenResponse,
    });

    const result = await drupalkit.simpleOauth.requestToken(
      "client_credentials",
      {
        client_id: CLIENT_ID,
        client_secret: CLIENT_SECRET,
      },
    );

    expect(result.ok).toBeTruthy();
  });

  it("should handle request errors", async () => {
    const drupalkit = createDrupalkit();

    mockResponse(fetchMock, drupalkit, {
      url: "/oauth/token",
      status: 400,
      payloadFixture: ErrorResponse,
    });

    const result = await drupalkit.simpleOauth.requestToken(
      "client_credentials",
      {
        client_id: CLIENT_ID,
        client_secret: CLIENT_SECRET,
      },
    );

    expect(result.err).toBeTruthy();

    if (result.err) {
      expect(result.val.statusCode).toEqual(400);
    }
  });

  it("should handle network errors when requesting token", async () => {
    const drupalkit = createDrupalkit();

    mockNetworkError(fetchMock);

    const result = await drupalkit.simpleOauth.requestToken(
      "client_credentials",
      {
        client_id: CLIENT_ID,
        client_secret: CLIENT_SECRET,
      },
    );

    expect(result.err).toBeTruthy();
  });
});
