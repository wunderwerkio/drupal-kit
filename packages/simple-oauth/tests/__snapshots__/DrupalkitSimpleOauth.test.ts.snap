// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DrupalkitSimpleOauth should request token with client credentials grant: request-payload 1`] = `"grant_type=client_credentials&client_id=12345678901234567890123456789012&client_secret=F9w1cM0GQw7GjjQUaZcscWHtxnMOvn4d"`;

exports[`DrupalkitSimpleOauth should request token with client credentials grant: token-response 1`] = `
{
  "access_token": "eyJ0eXAiOiJKV1QiLCJhbGciOiJSUzI1NiIsImp0aSI6IjZlN2ZiNmJiODNlNzZkOWIxOGU2MjAxZmZhYmY3YWFlZTE2OTZiOTMxZGFjOGQ0YTI4ODY0ZjJlZTI1ODA1MGYyZDAwZDE5NDYwNDAxMjY2In0.eyJhdWQiOiJtdm5wb2dJWU9zSngyQWw4YUdaemtIeTYzU3JJYVVvLU16d0pBSzhWU0lNIiwianRpIjoiNmU3ZmI2YmI4M2U3NmQ5YjE4ZTYyMDFmZmFiZjdhYWVlMTY5NmI5MzFkYWM4ZDRhMjg4NjRmMmVlMjU4MDUwZjJkMDBkMTk0NjA0MDEyNjYiLCJpYXQiOjE2Nzg4NjYzODksIm5iZiI6MTY3ODg2NjM4OSwiZXhwIjoxNjc4ODY2Njg5Ljk2MjIzMzEsInNjb3BlIjpbImFsbG93X2FsbCJdLCJzdWIiOiI0NSJ9.A6Ur4pciXtNSEl8mWEjW1DUll-zR4wSXjnS4FMnDHIV_mWclAQxgHuUexSYWnDASCpKC9TohqtSa2aPB_ymP2JhSaz9FTtYklFR2CmYz01HgZ83GGisNhvbdCx67CaWfluVS1fT-PFmut6fAUhbDWcQ7pt3sAZR4Rc6-iCWWe-lx5fnrli5S_egzuqJGj-NkoDg9s3yU6yvsPT8dObN-5Bl2qfU9VA21PYuWo89e-rm56mLexiuGUtPZHOa1hXGQ3WPEu3qlThtxaFgPyN4mFA0ql7JLV4Md2DrMcU8U598epro41KH9K6OvxOKbGFkQupXE-lODN0f3E4O_6Mr0Xk7YNtAna117fQEuoeAusJb024aZ-pxKNN2xjFUbesa6XvCNnkNKgDCaXWL4RWngD7lftYncsH6JoocqRgMmdzZGlf1A4fZtqaFs_DdC_JjjX2eyZZqVLv-wuJ04iGm3BXblUfBjaVx96tlhF30rfih9Cyu3psCzzoE7JLbqeAL6Vn6hQdZmZOyV_36-pOWEdZopy5tY2pZAmUAuUJFG40VEAGzh3dCuSVmni16f0OgQDS_ZpV0xCwEIc_UpkemY33EPVGU8zWWtztELVY6mAzEsR8xZwmtL8x4HadkU4wKoemos1KwwDM26tLR_x6K3H38aMl-M1PuCp2xkQg1gF_s",
  "expires_in": 300,
  "refresh_token": "def50200ca9707211ac5eee68c5acf9a84f169e0655580b3cfbadf479465d3cb26923cafdc827f459429b6445168c81903df56d2005df9212b63b712158a4b3e59db82c894bc9c88571e44dc8366b9951c5e11799b6b18fbb1eca8774986a6a161a8c4bebd6c604958bd123a3daef60adc2dd9e7acbf1c6003a5d1357a7c325da9b5123a8db4b1eed60abe7697019833359029d824599c799d5cd44ff4a21b3dc1f63b1ab96b958b77d91c2fde6c856762c47228cade0a2847c6c52abdd40ba960e1a53be4e5096489e73ecf143050c33c8ed63bd49d73ac56bf323fa970e77ba85faadf5896311cef5e2f3496f2a9851b48041d3b66227438f347e41726ac914467fd4f671b949f30c597a251e31b54811ab134f937f0df60c4b99819eb621fb3d110f07005354ea459e64b05d4642fa784d002b900047f63d8f55ca7cbbd4b1398e0ca1a6f2b76a1ec43a21072c4c3ba46846e7ae643682a7d8f4fa601de9febc854dd7f799323fccd4c4e94b5351022a6acbcb91462063707df028f9af39936ffe3a2a6127f5ce39b31260ff11c21fb43a8cd91fe4dff75",
  "token_type": "Bearer",
}
`;
